### PR TITLE
Make DisconnectMessage reason optional

### PR DESCRIPTION
**Pull request by mbaxter**
## PR description
* Update `DisconnectMessage` to hold an optional reason
* Add tests for `DisconnectMessage` construction and parsing
* Send an empty reason when we disconnect from blacklisted peers

### CODE REVIEW

File: ethereum/mock-p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/testing/MockNetwork.java [hunk 1]
```diff
@@ -246,13 +247,14 @@ public PeerInfo getPeer() {
...
-    public void terminateConnection(final DisconnectReason reason, final boolean peerInitiated) {
+    public void terminateConnection(
+        final Optional<DisconnectReason> reason, final boolean peerInitiated) {
       network.disconnect(this, reason);
     }
 
     @Override
     public void disconnect(final DisconnectReason reason) {
-      network.disconnect(this, reason);
+      network.disconnect(this, Optional.ofNullable(reason));
```
Comment (shemnon) [inline ethereum/mock-p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/testing/MockNetwork.java]: Should we push Optional<DisconnectReason> into PeerConnection?  An alternative would be to make a no-args disconnect() and make disconnect(@NotNull final DisconnectReason reason) and handle the optional stuff in the impls.

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/Callbacks.java [hunk 2]
```diff
@@ -44,12 +45,20 @@
...
       final PeerConnection connection,
-      final DisconnectReason reason,
+      final Optional<DisconnectReason> reason,
       final boolean initatedByPeer) {
     disconnectCallbacks.forEach(
         consumer -> consumer.onDisconnect(connection, reason, initatedByPeer));
   }
 
+  public void invokeDisconnect(
+      final PeerConnection connection,
```
Comment (shemnon) [inline ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/Callbacks.java]: Any reason we need both optional and non-optional forms?

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/transactions/TestNodeList.java [hunk 3]
```diff
@@ -218,7 +219,7 @@ public void assertNoNetworkDisconnections() {
     int errCnt = 0;
     final StringBuilder sb = new StringBuilder();
     for (final TestNode node : nodes) {
-      for (final Map.Entry<PeerConnection, DisconnectReason> entry :
+      for (final Map.Entry<PeerConnection, Optional<DisconnectReason>> entry :
```
[the remark this answers is not in the document]
Comment (mbaxter) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/transactions/TestNodeList.java]: 👍 Fixed

File: ethereum/mock-p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/testing/MockNetwork.java (hunk 1 above)
Comment (mbaxter) [reply ethereum/mock-p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/testing/MockNetwork.java]: I cleaned this up a little, but I have `disconnect()` and `disconnect(final DisconnectReason reason)` methods on `PeerConnection`.  I thought this was cleaner than having to wrap the reasons in `Optional`'s all over.

File: ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/Callbacks.java (hunk 2 above)
Comment (mbaxter) [reply ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/Callbacks.java]: Just for convenience - there are some cases where this api is called with a reason directly, and some cases where you're passing on the reason that's already wrapped in an `Optional`.

**Comment by mbaxter** [comment]
@shemnon - I think https://github.com/PegaSysEng/pantheon/pull/359 turned out better - what do you think?

**Comment by shemnon** [comment]
> @shemnon - I think #359 turned out better - what do you think?

#359 is much cleaner

**Comment by mbaxter** [comment]
Closing this in favor of #359

**Pull request closed without merging**